yb-voyager/src/queryparser/query_parser.go

```diff
@@ -163,4 +163,74 @@ func getCreateViewNode(parseTree *pg_query.ParseResult) (*pg_query.Node_ViewStmt
 func getCreateFuncStmtNode(parseTree *pg_query.ParseResult) (*pg_query.Node_CreateFunctionStmt, bool) {
 	node, ok := parseTree.Stmts[0].Stmt.Node.(*pg_query.Node_CreateFunctionStmt)
 	return node, ok
-}
\ No newline at end of file
+}
+
+func IsFunctionObject(parseTree *pg_query.ParseResult) bool {
+	funcNode, ok := getCreateFuncStmtNode(parseTree)
+	if !ok {
+		return false
+	}
+	return !funcNode.CreateFunctionStmt.IsProcedure
+}
+
+/*
+return type ex-
+CREATE OR REPLACE FUNCTION public.process_combined_tbl(
+    ... 
+)
+RETURNS public.combined_tbl.maddr%TYPE AS  
+return_type:{names:{string:{sval:"public"}}  names:{string:{sval:"combined_tbl"}}  names:{string:{sval:"maddr"}} 
+pct_type:true  typemod:-1  location:226}
+*/
+func GetReturnTypeOfFunc(parseTree *pg_query.ParseResult) string {
+	funcNode, _ := getCreateFuncStmtNode(parseTree)
+	returnType := funcNode.CreateFunctionStmt.GetReturnType()
+	return getParserTypeInString(returnType)
+}
+
+func getQualifiedTypeName(typeNames []*pg_query.Node) string {
+	var typeNameStrings []string
+	for _, n := range typeNames {
+		typeNameStrings = append(typeNameStrings, n.GetString_().Sval)
+	}
+	return strings.Join(typeNameStrings, ".")
+}
+
+func getParserTypeInString(typeVar *pg_query.TypeName) string {
+	typeNames := typeVar.GetNames()
+	finalTypeName := getQualifiedTypeName(typeNames) // type name can qualified table_name.column in case of %TYPE
+	if typeVar.PctType {                             // %TYPE declaration, so adding %TYPE for using it further
+		return finalTypeName + "%TYPE"
+	}
+	return finalTypeName
+}
+
+/*
+function ex - 
+CREATE OR REPLACE FUNCTION public.process_combined_tbl(
+    p_id int,        
+    p_c public.combined_tbl.c%TYPE,          
+    p_bitt public.combined_tbl.bitt%TYPE, 
+	..
+)
+parseTree-
+parameters:{function_parameter:{name:"p_id"  arg_type:{names:{string:{sval:"pg_catalog"}}  names:{string:{sval:"int4"}}  typemod:-1  location:66} 
+mode:FUNC_PARAM_DEFAULT}}  parameters:{function_parameter:{name:"p_c"  arg_type:{names:{string:{sval:"public"}}  names:{string:{sval:"combined_tbl"}}
+names:{string:{sval:"c"}}  pct_type:true  typemod:-1  location:87}  mode:FUNC_PARAM_DEFAULT}}  parameters:{function_parameter:{name:"p_bitt"  
+arg_type:{names:{string:{sval:"public"}}  names:{string:{sval:"combined_tbl"}}  names:{string:{sval:"bitt"}}  pct_type:true  typemod:-1  
+location:136}  mode:FUNC_PARAM_DEFAULT}} 
+
+*/
+func GetFuncParametersTypeNames(parseTree *pg_query.ParseResult) []string {
```

Comment:
(you can pick this up later since you are already refactoring...)
In-line with how you are refactoring currently, you could 
 - parse the function into a function DDLObject, 
- store all the types in parameters, return, etc, along with additional field `isPctType`. 

Then your detection can just go over and filter that list. Adding `%TYPE` suffix in `getParserTypeInString` isn't the cleanest way of doing it...